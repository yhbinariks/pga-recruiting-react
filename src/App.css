.app{
  padding: 10px;
}

table {
  font-family: arial, sans-serif;
  border-collapse: collapse;
}

th {
  cursor: pointer;
}

td, th {
  border: 1px solid #dddddd;
  text-align: left;
  padding: 8px;
}

tr:nth-child(even) {
  background-color: #dddddd;
}

.table__action-btn{
  cursor: pointer;
  color: blue;
  margin: 0 5px;
}

.player-form{
  padding: 10px 0 ;
}

.player-form__label{
  display: block;
  margin-bottom: 8px;
}
.player-form__label-text{
  display: inline-block;
  width: 100px;
}

.player-form__submit{
  padding: 0.4em;
  margin: 0 0 0.5em 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 2px;
  background-color: #f4f4f4;
  outline: none;
  text-transform: uppercase;
}
