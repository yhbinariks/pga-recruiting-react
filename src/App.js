import React, { Component } from 'react';
import Table from './components/Table/Table';
import {
  addId,
  parsePlayer,
  defaultPlayers,
  compareByNumField,
  compareByStringField,
} from './utils';
import Form from './components/Form';
import './App.css';

class App extends Component {
    state = {
      mode: 'add',
      playerId: null,
      players: defaultPlayers.map(addId),
      sortBy: 'score',
      isAsc: true,
    }

    componentDidMount() {
      this.sortPlayers();
    }

    sortPlayers = () => {
      const { sortBy, players, isAsc } = this.state;
      const sortedPlayers = players.slice().sort((p1, p2) => {
        const p1FullName = `${p1.lastName}${p1.firstName}`;
        const p2FullName = `${p2.lastName}${p2.firstName}`;
        if (sortBy === 'score') {
          if (p1.score === p2.score) return compareByStringField(p1FullName, p2FullName, isAsc);
          return compareByNumField(p1.score, p2.score, isAsc);
        }
        if (sortBy === 'name') {
          if (p1FullName === p2FullName) return compareByNumField(p1.score, p2.score, isAsc);
          return compareByStringField(p1FullName, p2FullName, isAsc);
        }

        if (p1[sortBy] === p2[sortBy]) return compareByStringField(p1FullName, p2FullName, isAsc);
        return compareByStringField(p1[sortBy].toString(), p2[sortBy].toString(), isAsc);
      });
      this.setState({
        players: sortedPlayers,
      });
    }

    onAdd = (event) => {
      const player = parsePlayer(Array.from(new FormData(event.target).entries()));
      this.setState(({ players }) => ({ players: [...players, player] }), this.sortPlayers);
    }

    setSortBy = sortBy => this.setState(prevState => ({
      sortBy,
      isAsc: prevState.sortBy === sortBy ? !prevState.isAsc : false,
    }), this.sortPlayers)

    onEdit = (event) => {
      const { players, playerId } = this.state;
      const player = parsePlayer(Array.from(new FormData(event.target).entries()));
      const newPlayers = players.map((p) => {
        if (p.id === playerId) return player;
        return p;
      });
      this.setState({
        mode: 'add',
        players: newPlayers,
        playerId: null,
      }, this.sortPlayers);
    }

    changeMode = (mode, playerId) => () => this.setState({ mode, playerId })

    onDelete = playerId => () => this.setState(prevState => ({
      mode: 'add',
      players: prevState.players.filter(player => player.id !== playerId),
    }), this.sortPlayers())

    onSubmit = (event) => {
      event.preventDefault();
      const { mode } = this.state;
      switch (mode) {
        case 'add': {
          return this.onAdd(event);
        }
        case 'edit': {
          return this.onEdit(event);
        }
        default:
          return null;
      }
    }

    render() {
      const { players, mode, playerId } = this.state;
      return (
        <main className="app">
          <Table
            players={players}
            changeMode={this.changeMode}
            onDelete={this.onDelete}
            setSortBy={this.setSortBy}
          />
          <Form
            key={playerId}
            player={players.find(player => player.id === playerId)}
            action={mode}
            onSubmit={this.onSubmit}
          />
        </main>
      );
    }
}

export default App;
