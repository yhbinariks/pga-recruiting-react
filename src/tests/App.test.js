import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import App from '../App';

const mockMath = Object.create(global.Math);
mockMath.random = () => 0.555555555555555555555555555555;
global.Math = mockMath;

describe('App component', () => {
  it('snapshot', () => {
    const app = shallow(<App />);
    expect(toJson(app)).toMatchSnapshot();
  });
});
