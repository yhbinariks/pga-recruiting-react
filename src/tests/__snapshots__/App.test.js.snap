// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component snapshot 1`] = `
<main
  className="app"
>
  <Table
    changeMode={[Function]}
    onDelete={[Function]}
    players={
      Array [
        Object {
          "firstName": "Rob",
          "id": "RobVera-k",
          "lastName": "Vera",
          "score": 88,
        },
        Object {
          "firstName": "Alice",
          "id": "AliceGeary-k",
          "lastName": "Geary",
          "score": 96,
        },
        Object {
          "firstName": "Bob",
          "id": "BobGeary-k",
          "lastName": "Geary",
          "score": 96,
        },
        Object {
          "firstName": "John",
          "id": "JohnJunge-k",
          "lastName": "Junge",
          "score": 96,
        },
      ]
    }
    setSortBy={[Function]}
  />
  <Form
    action="add"
    key="null"
    onSubmit={[Function]}
    player={
      Object {
        "firstName": "",
        "lastName": "",
        "score": 0,
      }
    }
  />
</main>
`;
